Add --log-wptreport option to wptrunner

sideshowbarker opened the pull request:
This change adds a --log-wptreport option to wptrunner; the option
enables wptrunner to output test results in the JSON format that
wpreport[1] expects to consume as input.

[1] https://www.npmjs.com/package/wptreport

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/w3c/web-platform-tests/6356)
<!-- Reviewable:end -->

sideshowbarker committed:
Add --log-wptreport option to wptrunner

This change adds a --log-wptreport option to wptrunner; the option
enables wptrunner to output test results in the JSON format that
wpreport[1] expects to consume as input.

[1] https://www.npmjs.com/package/wptreport

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/247884179)

# Firefox (nightly) #
Testing web-platform-tests at revision 310bb95a398bd98f39e24f48e515bee30fef277b
Using browser at version BuildID 20170612100241; SourceStamp 27cad9749cddf68e11fdd4e5d73dad84a8f8cf23
Starting 10 test iterations
No tests run.

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/247884180)

# Sauce (safari) #
Testing web-platform-tests at revision 310bb95a398bd98f39e24f48e515bee30fef277b
Using browser at version 10.0
Starting 10 test iterations
No tests run.

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/247884186)

# Chrome (unstable) #
Testing web-platform-tests at revision 310bb95a398bd98f39e24f48e515bee30fef277b

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/247884187)

# Sauce (MicrosoftEdge) #
Testing web-platform-tests at revision 310bb95a398bd98f39e24f48e515bee30fef277b

bobholt commented on tools/wptrunner/wptrunner/formatters.py:
```diff
@@ -0,0 +1,66 @@
+        subtest["status"] = data["status"]
+        if data.get("message"):
+            if subtest["message"]:
+                subtest["message"] += ", " + data["message"]
```
Would `subtest["message"]` be better as a `list`? We use a `set` in `check_stability.py` to avoid duplicates. That's not JSON-serializable, so a `list` could work in a pinch.

sideshowbarker replied:
> Would `subtest["message"]` be better as a `list`? We use a `set` in `check_stability.py` to avoid duplicates. That's not JSON-serializable, so a `list` could work in a pinch.

Yeah but per @jgraham’s comment at https://reviewable.io/reviews/w3c/web-platform-tests/6356#-KnjvOnr7eDxNtM6O8Ml:-KnjvOnr7eDxNtM6O8Mm:b-35gses I’ve now made it just overwrite any existing data for a subtest if there’s one with the same name

bobholt commented:
This is probably a larger discussion that doesn't impact whether this PR should be merged, but it looks similar enough to the formatting `check_stability` does (https://github.com/w3c/web-platform-tests/blob/master/check_stability.py#L616-L664) to make me wonder:

If we are going to add a formatter option to `wptcommandline`, could we consider how it could replace the log formatting/handling logic in `check_stability.py`? `check_stability` invokes `wptcommandline` to check its args, but otherwise runs `wptrunner` directly. If `wptcommandline` and could otherwise do the custom things `check_stability` needs to do with `wptrunner`, we could remove a lot of formatting gymnastics from `check_stability`.

Doing that could also reduce the [reformatting I have to do](https://github.com/bobholt/web-platform-tests/pull/22/files#diff-6fbabfc979e734908cfe751497cd42c5R801) to get the OrderedDict/Set format check_stability currently uses into a serializable format for posting to an API endpoint.

gsnedders commented:
Also related to #5900.

jgraham commented:
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[tools/wptrunner/wptrunner/formatters.py, line 11 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/6356#-KnjtfxbRvjUIZsfL3S_:-KnjtfxcOHPKsa54_tb9:bntup0p) ([raw file](https://github.com/w3c/web-platform-tests/blob/39b5dc35d18e0ca34d1e20870667b4d842f87872/tools/wptrunner/wptrunner/formatters.py#L11)):*
> ```Python
> 
>     def __init__(self):
>         self.raw_results = OrderedDict()
> ```

Is there any point in using `OrderedDict` here (and elsewhere)? Presumably the order is non-deterministic on reading the created json.

---

*[tools/wptrunner/wptrunner/formatters.py, line 23 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/6356#-KnjuEoP6OKm6NN6EZOX:-KnjuEoQpjoBWsNxPxRs:b-tfzlgp) ([raw file](https://github.com/w3c/web-platform-tests/blob/39b5dc35d18e0ca34d1e20870667b4d842f87872/tools/wptrunner/wptrunner/formatters.py#L23)):*
> ```Python
> 
>     def test_start(self, data):
>         self.start_times[data["test"]] = data["time"]
> ```

Does this exist? Is it ever used?

---

*[tools/wptrunner/wptrunner/formatters.py, line 27 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/6356#-Knjueu7QDK5ag55Qi_Y:-Knjueu7QDK5ag55Qi_Z:b3fu3hh) ([raw file](https://github.com/w3c/web-platform-tests/blob/39b5dc35d18e0ca34d1e20870667b4d842f87872/tools/wptrunner/wptrunner/formatters.py#L27)):*
> ```Python
>     def find_or_create_test(self, data):
>         test_name = data["test"]
>         if self.raw_results.get(test_name):
> ```

I guess I would write this as

```
if test_name not in self.raw_results:
    self.raw_results[test_name] = {"subtests": [], "status": "", message: None}
return self.raw_results[test_name]
```

but it's not super important.

---

*[tools/wptrunner/wptrunner/formatters.py, line 58 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/6356#-KnjvOnr7eDxNtM6O8Ml:-KnjvOnr7eDxNtM6O8Mm:b-35gses) ([raw file](https://github.com/w3c/web-platform-tests/blob/39b5dc35d18e0ca34d1e20870667b4d842f87872/tools/wptrunner/wptrunner/formatters.py#L58)):*
> ```Python
>         subtest = self.find_or_create_subtest(data)
>         subtest["status"] = data["status"]
>         if data.get("message"):
> ```

So theoretically getting the same subtest name is an error. In practice we probably keep the last one rather than concataneting the messages together and taking the last status.

---

*[tools/wptrunner/wptrunner/formatters.py, line 66 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/6356#-KnjvWPd6Ux4pYyLUiTx:-KnjvWPeS92-l9doOEY2:btjhijg) ([raw file](https://github.com/w3c/web-platform-tests/blob/39b5dc35d18e0ca34d1e20870667b4d842f87872/tools/wptrunner/wptrunner/formatters.py#L66)):*
> ```Python
>     def test_end(self, data):
>         test = self.find_or_create_test(data)
>         test["status"] = data["status"]
> ```

Shouldn't this also set the message, if any?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/web-platform-tests/6356)*
<!-- Sent from Reviewable.io -->

jgraham commented:
> If we are going to add a formatter option to wptcommandline, could we consider how it could replace the log formatting/handling logic in check_stability.py?

I assume that the goal of making this work for multiple runs would conflict with the goal of making this compatible with the existing node module, otherwise it could be a good idea.

sideshowbarker committed:
fixup! Add --log-wptreport option to wptrunner

sideshowbarker commented:
Review status: 1 of 2 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[tools/wptrunner/wptrunner/formatters.py, line 11 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/6356#-KnjtfxbRvjUIZsfL3S_:-KoH10Zofk8PpkvGKsDI:bctbo2g) ([raw file](https://github.com/w3c/web-platform-tests/blob/39b5dc35d18e0ca34d1e20870667b4d842f87872/tools/wptrunner/wptrunner/formatters.py#L11)):*
> Is there any point in using `OrderedDict` here (and elsewhere)? Presumably the order is non-deterministic on reading the created json.

No point; changed it to just normal dictionary

---

*[tools/wptrunner/wptrunner/formatters.py, line 23 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/6356#-KnjuEoP6OKm6NN6EZOX:-KoH1hWg26Ykt6ksIaU_:bcpfx3r) ([raw file](https://github.com/w3c/web-platform-tests/blob/39b5dc35d18e0ca34d1e20870667b4d842f87872/tools/wptrunner/wptrunner/formatters.py#L23)):*
>> def test_start(self, data)
>
> Does this exist? Is it ever used?

Was not used; no need for it, removed

---

*[tools/wptrunner/wptrunner/formatters.py, line 27 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/6356#-Knjueu7QDK5ag55Qi_Y:-KoH1sJrHsilWbyYhCtD:b-vpy7eb) ([raw file](https://github.com/w3c/web-platform-tests/blob/39b5dc35d18e0ca34d1e20870667b4d842f87872/tools/wptrunner/wptrunner/formatters.py#L27)):*
> I guess I would write this as…

Thanks yeah definitely an improvement, changed it to your suggestion

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/web-platform-tests/6356)*
<!-- Sent from Reviewable.io -->

sideshowbarker committed:
fixup! Add --log-wptreport option to wptrunner

sideshowbarker commented:
Review status: 1 of 2 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[tools/wptrunner/wptrunner/formatters.py, line 58 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/6356#-KnjvOnr7eDxNtM6O8Ml:-KoH7ZncUw8kk_keA4-b:bs3ypn0) ([raw file](https://github.com/w3c/web-platform-tests/blob/39b5dc35d18e0ca34d1e20870667b4d842f87872/tools/wptrunner/wptrunner/formatters.py#L58)):*
> So theoretically getting the same subtest name is an error. In practice we probably keep the last one rather than concataneting the messages together and taking the last status.

OK, I have change it so it just creates/overwrites a subtest with the name

---

*[tools/wptrunner/wptrunner/formatters.py, line 66 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/6356#-KnjvWPd6Ux4pYyLUiTx:-KoH7sh9DSE8PeJefdIj:bhhrz7m) ([raw file](https://github.com/w3c/web-platform-tests/blob/39b5dc35d18e0ca34d1e20870667b4d842f87872/tools/wptrunner/wptrunner/formatters.py#L66)):*
> ``` 
> test_end(self, data):
>         test = self.find_or_create_test(data)
>         test["status"] = data["status"]
> ```
> 
> Shouldn't this also set the message, if any?

Yes, have made it now do that

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/web-platform-tests/6356)*
<!-- Sent from Reviewable.io -->

sideshowbarker commented:
> If we are going to add a formatter option to `wptcommandline`, could we consider how it could replace the log formatting/handling logic in `check_stability.py`?

Yes, I think we should

> I assume that the goal of making this work for multiple runs would conflict with the goal of making this compatible with the existing node module, otherwise it could be a good idea.

It need not conflict. If somebody takes output from multiple runs and feeds it to the existing `wpreport` node module, I think it will (should) just fail and that’s fine. The solution to that problem is just that they should not feed output from multiple runs to `wpreport`, because by design it’s intended to only consume and report on the results of a single run.

> Also related to #5900 

Yes, so we should also rope in @jeffcarp to look this over

sideshowbarker commented:
By the way, if we decide to use this report formatter for the #5900 need and for the `check_stability.py` then I think we also need to consider renaming the option (and underlying class in the code) to something more general instead of `wptreport`. Not sure what to suggest though…

jgraham commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/web-platform-tests/6356)*
<!-- Sent from Reviewable.io -->

jeffcarp commented:
Lgtm! The format of this output contains all the things we need, so #5900 should no longer be necessary.

With respect to the flag name, IMHO `wptreport` sounds general enough - if it's understood to mean "the format `wptreport` expects (whose structure is agreed upon and won't change)".
